core/trace/src/main/scala/org/typelevel/otel4s/trace/TraceState.scala

```diff
@@ -0,0 +1,254 @@
+  *     a lowercase letter, and can only contain lowercase letters a-z, digits
+  *     0-9, underscores _, dashes -, asterisks *, and forward slashes /.
+  *
+  *   - Value is opaque string up to 256 characters printable ASCII RFC0020
+  *     characters (i.e., the range 0x20 to 0x7E) except comma , and =.
+  *
+  * @see
+  *   [[https://opentelemetry.io/docs/specs/otel/trace/api/#tracestate]]
+  *   [[https://www.w3.org/TR/trace-context/#mutating-the-tracestate-field]]
+  *   [[https://www.w3.org/TR/trace-context/#tracestate-header-field-values]]
+  */
+sealed trait TraceState {
+
+  /** Returns the value to which the specified key is mapped, or `None` if this
+    * map contains no mapping for the key.
+    *
+    * @param key
+    *   with which the specified value is to be associated
+    */
+  def get(key: String): Option[String]
+
+  /** Adds or updates the entry that has the given `key` if it is present. If
+    * either the key or the value is invalid, the entry will not be added.
+    *
+    * @param key
+    *   the key for the entry to be added. Key is an opaque string up to 256
+    *   characters printable. It MUST begin with a lowercase letter, and can
+    *   only contain lowercase letters a-z, digits 0-9, underscores _, dashes -,
+    *   asterisks *, and forward slashes /. For multi-tenant vendor scenarios,
+    *   an at sign (@) can be used to prefix the vendor name. The tenant id
+    *   (before the '@') is limited to 240 characters and the vendor id is
+    *   limited to 13 characters. If in the multi-tenant vendor format, then the
+    *   first character may additionally be numeric.
+    *
+    * @param value
+    *   the value for the entry to be added. Value is opaque string up to 256
+    *   characters printable ASCII RFC0020 characters (i.e., the range 0x20 to
+    *   0x7E) except comma , and =.
+    *
+    * @return
+    *   a new instance of [[TraceState]] with added entry
+    */
+  def updated(key: String, value: String): TraceState
+
+  /** Removes the entry that has the given `key` if it is present.
+    *
+    * @param key
+    *   the key for the entry to be removed.
+    *
+    * @return
+    *   a new instance of [[TraceState]] with removed entry
+    */
+  def removed(key: String): TraceState
+
+  /** Returns the number of entries in this state. */
+  def size: Int
+
+  /** Returns whether this state is empty, containing no entries. */
+  def isEmpty: Boolean
+
+  /** Returns a map representation of this state. */
+  def asMap: SeqMap[String, String]
+
+  override final def hashCode(): Int =
+    Hash[TraceState].hash(this)
+
+  override final def equals(obj: Any): Boolean =
+    obj match {
+      case other: TraceState => Hash[TraceState].eqv(this, other)
+      case _                 => false
+    }
+
+  override final def toString: String =
+    Show[TraceState].show(this)
+}
+
+object TraceState {
+
+  /** An empty [[TraceState]]. */
+  val Empty: TraceState = MapBasedTraceState(Vector.empty)
```

Comment:
nitpick: the stdlib tends to use lowercase for `empty`
```suggestion
  val empty: TraceState = MapBasedTraceState(Vector.empty)
```

Reply:
Should it be defined as:
```scala
private val Empty: TraceState = MapBasedTraceState(Vector.empty)

def empty: TraceState = Empty
```
?

It could be slightly better in terms of bincompat management. But I'm fine with `val empty: TraceState = MapBased...` too.

Reply:
> It could be slightly better in terms of bincompat management

I don't think it makes a difference, but it would be better for source-compat management.